### PR TITLE
Debt coupon manager

### PR DESCRIPTION
**Pull request by zgorizzo69**
:robot: **What has been done**
* refactor UAD to concentrate all access control inside the UAD Manager
* update of UAD Manager with new roles and contract dependencies
* full debtCoupon issuance and redeem flow tests
* fix  dollar minting calculator + unit tests
* update excess dollar distributor + unit tests
* Add ABDK libs to do precision decimal calculus
* changed license to MIT
* Add logic for upgradable hook incentive 
* add new mock for testing purposes
* add hardhat mine N block utility function

:nerd_face: **What needs to be done** - _in the next PR_
* add more tests for debtCouponManager
* add logic and tests to incentive

### CODE REVIEW

File: contracts/ExcessDollarsDistributor.sol [hunk 1]
```diff
@@ -1,14 +1,17 @@
-// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
 pragma solidity ^0.8.3;
 
 import "hardhat/console.sol";
```
Comment (zapaz) [inline contracts/ExcessDollarsDistributor.sol:4]: can be removed

File: contracts/interfaces/ISablier.sol [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
```
Comment (zapaz) [inline contracts/interfaces/ISablier.sol:1]: same license for external lib or not ?

File: contracts/DollarMintingCalculator.sol [hunk 3]
```diff
@@ -18,9 +22,19 @@ contract DollarMintingCalculator is IDollarMintingCalculator {
     function getDollarsToMint() external view override returns (uint256) {
         TWAPOracle oracle = TWAPOracle(manager.twapOracleAddress());
         uint256 twapPrice = oracle.consult(manager.uADTokenAddress());
+        bytes16 one = (uint256(1 ether)).fromUInt();
```
Comment (zapaz) [inline contracts/DollarMintingCalculator.sol:25]: should add "require" twapPrice > one  

(currently revert in the case twapPrice < one  without reason message, due to toUint of a negative number)

File: test/DollarMintingCalculator.test.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    const uadPrice = ethers.utils.parseEther("1.054678911145683254");
+    await setup(totSupply, uadPrice, uadPrice);
+    // check tfor overflow revert
+    const toMint = await dollarMintingCalculator.getDollarsToMint();
+    const calculatedToMint = calcDollarsToMint(
+      new Big(totSupply.toString()),
+      new Big(uadPrice.toString())
+    );
+    expect(toMint).not.to.equal(calculatedToMint);
```
Comment (zapaz) [inline test/DollarMintingCalculator.test.ts:126]: should test with some precision p =>   delta < 10**-p

(as done for tests of durationMultiplier in BondingShare branch)

**Review by zapaz — COMMENTED** [review]
ABDK calculation review

File: contracts/interfaces/ISablier.sol (hunk 2 above)
Comment (zgorizzo69) [reply contracts/interfaces/ISablier.sol:1]: right now we will not touch the ext libs

File: contracts/ExcessDollarsDistributor.sol (hunk 1 above)
Comment (zgorizzo69) [reply contracts/ExcessDollarsDistributor.sol:4]: done

File: contracts/DollarMintingCalculator.sol (hunk 3 above)
Comment (zgorizzo69) [reply contracts/DollarMintingCalculator.sol:25]: done

**Review by zapaz — APPROVED** [review]

**Pull request merged**